src/coreclr/gc/gc.cpp

```diff
@@ -12093,30 +12335,52 @@ void gc_heap::distribute_free_regions()
         // just to reduce the number of #ifdefs in the code below
         const int i = 0;
 #endif //MULTIPLE_HEAPS
+
         for (int kind = basic_free_region; kind < kind_count; kind++)
         {
-            total_num_free_regions[kind] += hp->free_regions[kind].get_num_free_regions();
+            // If there are regions in free that haven't been used in 20 GCs we always decommit them.
+            region_free_list& region_list = hp->free_regions[kind];
+            heap_segment* next_region = nullptr;
+            for (heap_segment* region = region_list.get_first_free_region(); region != nullptr; region = next_region)
+            {
+                next_region = heap_segment_next (region);
+                if (heap_segment_age_in_free (region) >= AGE_IN_FREE_TO_DECOMMIT)
+                {
+                    num_decommit_regions_by_time++;
+                    size_decommit_regions_by_time += get_region_committed_size (region);
+                    dprintf (REGIONS_LOG, ("h%2d region %Ix age %2d, decommit", 
+                        i, heap_segment_mem (region), heap_segment_age_in_free (region)));
+                    region_free_list::unlink_region (region);
+                    region_free_list::add_region (region, global_regions_to_decommit);
+                }
+            }
```

Comment:
what if suddenly we have a much bigger budget (this could happen if gen0's surv rate is suddenly very high due to something like a `ConcurrentDictionary` resize) and we'd need the ones with high ages? now we've moved them to the decommit list and we won't take them into consideration for filling budget.

Reply:
yep I saw that. what I meant was this - let's say all the regions on the decommit list were already decommitted so nothing on the decommit list when we started `distribute_free_regions`. and we have a bunch of regions whose age is `>= AGE_IN_FREE_TO_DECOMMIT`, and we are putting them on the decommit list but later in this method we don't actually try to move any regions from the decommit list if we don't have enough free regions to accommodate SOH budget. that's how I read the code, maybe I'm missing something?

Reply:
Ah, now I see what you mean.

In this situation there is a conflict between what the budget says (e.g. "we need 10 regions") and what the age says ("but there are 2 we haven't used in the last 20 GCs"). I was thinking about the situation where our gen 1 budget is way larger than what we actually use, and so decided to let the age take precedence.

This is likely the right thing to do in this situation with gen 1, but it doesn't address the situation where the budget suddenly got larger as you described above. I'll think about it...